### DISCUSSION TITLE
Using mu with birdbrain technologies hummingbird/finch robot

### DISCUSSION THREAD

**Discussion by fmorton** [mu-editor/mu#1344] [Q&A]
Birdbrain Technologies provides their library as a single file contain classes for their devices. The file is available at:

https://www.birdbraintechnologies.com/downloads/installers/BirdBrainPython.zip

To make this available in the Python 3 mode on a Mac, I copied the BirdBrain.py file into the site-packages directory buried under the Library/Application Support directory, specifically on my system:

~/Library/Application Support/mu/mu_venv/lib/python3.7/site-packages

Is this the best way to add a class library like this to sys.path?

Thanks.

**Comment by carlosperate** [comment]
Is that a "desktop" Python module? Is it meant for a Raspberry Pi? Or how does it interface with the robot?
If the code is in github it can probably be pip installed via URL, and that could probably be done via the Third Party packages window. Although I haven't tried that yet.

**Comment by dybber** [comment]
Perhaps off-topic, but I'll add it anyway: if the library is a single file, I would just ask my pupils/students to copy it to the directory of their project. I think that kids should also learn that there's nothing "magical" about creating libraries and showing them that it's just another Python file is a nice illustration of that.

As this is not really an issue, I will convert this to a discussion in our discussion forum until we've figured out if there's something that needs to change in Mu.

**Reply by carlosperate** [reply]
Thanks for transferring!
Maybe the outcome will be an update to the documentation.

**Comment by carlosperate** [comment]
I would have to check, but I believe the mu_code folder is added to the site-packages paths, so adding the package folder there should make it importable. Just need to make sure the folder structure is correct (it is not subnested in more directories than it needs).
If these are individual modules they can be added next to the user python code (like @dybber suggested) or they can probably be also added the mu_code folder.

**Comment by fmorton** [comment]
The BirdBrain library is not meant for the raspberry pi. They have two "robot" kind of devices, one called a Hummingbird (which you attach sensors and output devices to) and the Finch 2, which is a self-contained robot. On your system is a "BlueBird Connector" which appears like a web server on a non-standard port to your app running on your system. The "BlueBird Connector" then takes those http requests and uses BLE to communicate to the device. Very nice for writing your own drivers. Young students also do very well with this.

The BirdBrain Python library is just a single file with some class definitions, so there is no sort of packaging with the code.

On my system (Mac with Big Sur), the sys.path contains:

['/Users/fmorton/tmp', '/Users/fmorton/.pyenv/versions/3.7.6/lib/python37.zip', '/Users/fmorton/.pyenv/versions/3.7.6/lib/python3.7', '/Users/fmorton/.pyenv/versions/3.7.6/lib/python3.7/lib-dynload', '/Users/fmorton/Library/Application Support/mu/mu_venv/lib/python3.7/site-packages']

I do not see mu_code in the path. For this case, it would be nice to have mu_code/site-packages, but this is probably an outlier. It is fine for me to put it in the mu_env site-packages. I just wanted to make sure there was not some well-known place to put files like this that I did not know about.

I did try the third-party packages without success, but this library is not really a package. It is just a file with some class definitions.

I will also consider the @dybber advise about just adding the library file with the project. One challenge with that is I plan to make changes to the library and I want to avoid old versions floating around. I don't actually copy it into the site-packages directory, but instead put a symbolic link pointing to the github repository on the systems that we put in a standardized place, so everyone is always using the latest library.

BTW, the mu editor is absolutely fantastic with the birdbrain devices. You just use the Python 3 mode and everything works with almost no effort. Well done!

**Reply by carlosperate** [reply]
> I do not see mu_code in the path. 

Yeah, sorry, maybe this was a feature discussed at some point but not really implemented (or something we had before and has been removed) so only code run from the `mu_code` directory would have that folder added to the path (as normal from running any python file).

> I did try the third-party packages without success, but this library is not really a package. It is just a file with some class definitions.

Yeah, unfortunately those modules don't look packaged at all, just zipped together. In that case the simplest option is to place them next to the user code, and that should work with any Python interpreter.

> One challenge with that is I plan to make changes to the library and I want to avoid old versions floating around.

Yeah, that is a fair request. Maybe having a `site_packages` folder inside `mu_code` could be the proposal from this discussion.

>BTW, the mu editor is absolutely fantastic with the birdbrain devices. You just use the Python 3 mode and everything works with almost no effort. Well done!

I'm glad to hear that!
